[WPE][IPC] Add support for AHardwareBuffer attachments

aperezdc opened the pull request:
#### 17242045db7fe32a02b655bc64304f3eb157099c
<pre>
[WPE][IPC] Add support for AHardwareBuffer attachments
<a href="https://bugs.webkit.org/show_bug.cgi?id=297592">https://bugs.webkit.org/show_bug.cgi?id=297592</a>

Reviewed by Carlos Garcia Campos.

When building for Android, make message attachments a variant to
allow passing both UnixFileDescriptor or AHardwareBuffer through
IPC. This involved the following related changes in how Attachment
objects are handled: a new AttachmentInfo::m_type member indicates
whether an attachment is an UnixFileDescriptor or an AHardwareBuffer,
and alternative code paths are needed used to act them.

The trickier part of actually sending and receiving AHardwareBuffer
objects: we cannot &quot;open up&quot; the inside of buffer, and passing them
around needs to be done with AHardwareBuffer_sendHandleToUnixSocket()
and AHardwareBuffer_recvHandleFromUnixSocket(). These function will
themselves send/receive a message for each AHardwareBuffer, out of
our control. Therefore, on the writing end:

  - When preparing to send a message, note the amount of buffer
    handles in a new MessageInfo::m_validHardwareBufferCount
    member (invalid/null handles are not sent).

  - Store the valid AHardwareBuffer handles to be sent later and
    send the message (sans buffer handles) in the usual fashion.

  - Send the AHardwareBuffer handles right after the message.
    At the end of Connection::sendOutputMessage(), store the buffer
    handles Connection::m_outgoingHardwareBuffers, and call
    ::sendOutgoingHardwareBuffers().

  - The ::sendOutgoingHardwareBuffers() function attempts to send
    as many handles as it can right away. If the operation would
    block, it makes use of the socket monitor to re-schedule calling
    itself when the socket is ready again for sending. This is
    repeated until there are no more buffers left.

On the receiving end things are a little bit trickier because the
socket monitor is always armed to invoke Connection::readyReadHandler()
and this means that we cannot start it independently to handle only
retiring AHardwareBuffer handles from the socket -- we need to thread
the code to retire buffer handles into the existing functions:

  - The ::readyReadHandler() function may be invoked either to
    retry sending the &quot;main&quot; message, or to retry sending buffer
    handles. The Connection::m_pendingIncomingHardwareBufferCount
    is used to determine which code path to follow.

  - First, if there are no pending buffer handles to receive, a
    message is read. At the moment the code assumes that message
    are always read in their entirety, which is the case with
    SOCK_SEQPACKET.

  - The read message up to this point includes both the payload and
    file descriptors. The MessageInfo in the header of the message
    indicates how many AHardwareBuffer handles we need to retire.
    This is saved in Connection::m_pendingIncomingHardwareBufferCount
    before calling ::receiveIncomingHardwareBuffers().

  - Then ::receiveIncomingHardwareBuffers() attempts retiring from
    the socket as many buffer handles as it can. If reading handles
    would block, the function returns immediately, and will rely on
    the next call to ::readyReadHandler() -- as the socket monitor
    is always kept armed and invoke it for reading.

  - When execution reaches ::readyReadHandler() again and, crucially,
    there are still AHardwareBuffer handles pending to be retired from
    the socket, it will call again ::receiveIncomingHardwareBuffers()
    to resume the task. When all the handles have been retired from
    the socket we still need a call to ::processMessage() at this point,
    but without going first through trying to read another complete
    message.

* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/android/RefPtrAndroid.cpp: Added.
(WTF::DefaultRefDerefTraits&lt;AHardwareBuffer&gt;::refIfNotNull):
(WTF::DefaultRefDerefTraits&lt;AHardwareBuffer&gt;::derefIfNotNull):
* Source/WTF/wtf/android/RefPtrAndroid.h: Added.
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/Attachment.h:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/android/ArgumentCodersAndroid.cpp: Added.
(IPC::ArgumentCoder&lt;UnixFileDescriptor&gt;::decode):
(IPC::ArgumentCoder&lt;RefPtr&lt;AHardwareBuffer&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;RefPtr&lt;AHardwareBuffer&gt;&gt;::decode):
* Source/WebKit/Platform/IPC/android/ArgumentCodersAndroid.h: Added.
* Source/WebKit/Platform/IPC/unix/ArgumentCodersUnix.cpp:
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::AttachmentInfo::AttachmentInfo):
(IPC::AttachmentInfo::type const):
(IPC::AttachmentInfo::setType):
(IPC::Connection::processMessage):
(IPC::Connection::readyReadHandler):
(IPC::Connection::platformCanSendOutgoingMessages const):
(IPC::Connection::sendOutputMessage):
(IPC::Connection::sendOutgoingHardwareBuffers):
(IPC::Connection::receiveIncomingHardwareBuffers):
* Source/WebKit/Platform/IPC/unix/UnixMessage.h:
(IPC::MessageInfo::MessageInfo):
(IPC::MessageInfo::hardwareBufferCount const):
(IPC::UnixMessage::UnixMessage):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesWPE.txt:

Canonical link: <a href="https://commits.webkit.org/299311@main">https://commits.webkit.org/299311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80b0c3c3e8e7eae66a0ebe31b28aabd6e661fd06

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118590 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38271 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28922 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124767 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70649 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38967 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46853 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/90006 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121543 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31035 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/106326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70510 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24437 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68424 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110708 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100473 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24628 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127828 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117104 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45497 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/34324 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98638 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45861 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102547 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98422 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25020 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43864 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21864 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45367 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51045 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145800 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44830 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37489 "Found 1 new JSC binary failure: testapi, Found 18606 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInPrimitive.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48177 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46517 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/Platform/IPC/unix/UnixMessage.h:
```diff
@@ -75,15 +87,30 @@ class MessageInfo {
     size_t m_bodySize;
     size_t m_attachmentCount;
+#if OS(ANDROID)
+    size_t m_validHardwareBufferCount;
```
If anyone is worried about growing this by 8 bytes, we could change the types to `uint8_t`/`uint16_t` because the maximum number of attachments is 254+1, there's the following constants in `ConnectionUnix.cpp`:

```cpp
static const size_t messageMaxSize = 4096;
static const size_t attachmentMaxAmount = 254;
```

I expect the compiler would still add some padding, but in most architectures each member would be 32-bit aligned and the size of the type would be 12 bytes (16 bytes on Android), instead of the current 24 bytes (32 bytes on Android).

carlosgcampos replied:
I wonder why we need this, though. In the case of file descriptors we check the amount of non-null fds on the receiver side, and fill m_fileDescriptors vector with the "valid" ones. We are also checking non-null buffers, so I guess we could do ther same without having to include this info in the message, no? I also wonder if it's possible to send null android buffers.

aperezdc replied:
The idea was to avoid yet one more traversal of the array of `AttachmentInfo`s. This is a hot code path after all.

Regarding null buffers, I *think* it is unlikely that we would need to pass them over IPC; but given that the type used everywhere is `RefPtr<AHardwareBuffer>`, which could be null, I decided to handle them in the IPC code.

I am not sure how to proceed. Would you suggest asserting that AHBs cannot be null when they reach the IPC code and simplifying a little bit? (although it will not be much of simplification).

aperezdc commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -468,10 +541,26 @@ bool Connection::sendOutputMessage(UnixMessage& outputMessage)
+                }
+            );
+        }
+#else
```
Something I don't like much of this function is that it will prepare the arrays of file descriptors, AHBs, and other minutiae every time the function is called. It would be interesting to know how often this is getting re-scheduled on `EAGAIN`/`EWOULDBLOCK` because it may be worth it to store all the “prepared” data (even including the `struct iov`, message header, and so on) as part of `Connection::m_pendingOutputMessage` and avoid re-calculating each time.

Anyway, this is just an idea that came to my mind while working on the Unix IPC code, for this PR it seems much better to keep the existing structure. Refactoring can be done later, if we think it's worth it.

carlosgcampos commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -42,8 +42,12 @@
+#endif
 #if USE(GLIB)
 #include "ArgumentCodersGlib.h"
+#include "RendererBufferFormat.h"
```
Why do we need this here?

aperezdc replied:
I needed this to fix the build. By chance the bots and x86_64 build works because of the compilation units picked for each unified build source 🙃

carlosgcampos commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -721,6 +725,18 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
     std::unique_ptr<UnixMessage> m_pendingOutputMessage;
+
+#if OS(ANDROID)
+    size_t m_pendingIncomingHardwareBufferCount;
```
```suggestion
    size_t m_pendingIncomingHardwareBufferCount { 0 };
```

carlosgcampos commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -721,6 +725,18 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
+    bool sendOutgoingHardwareBuffers();
+
+    bool receiveHardwareBuffers(size_t amount);
+    bool receiveIncomingHardwareBuffers();
```
Move the functions before the variables.

carlosgcampos commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -559,8 +664,104 @@ bool Connection::sendOutputMessage(UnixMessage& outputMessage)
+    RELEASE_ASSERT(m_outgoingHardwareBuffers.isEmpty());
+    m_outgoingHardwareBuffers = WTFMove(hardwareBuffers);
+    return sendOutgoingHardwareBuffers();
+}
```
Why do we need this function? I guess we could do this in sendOutputMessage since it's only called there.

aperezdc replied:
For me it helps understand better how the code works: one prepares to start the process, does checks, etc., and then calls the other that actually does it, and the second one is the only function that gets re-scheduled in case of `EAGAIN`/`EWOULDBLOCK`.

The above is only a personal preference, so I'll move this as suggested.

carlosgcampos commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -559,8 +664,104 @@ bool Connection::sendOutputMessage(UnixMessage& outputMessage)
+
+    m_pendingIncomingHardwareBufferCount = bufferCount;
+    return receiveIncomingHardwareBuffers();
+}
```
Same here

aperezdc replied:
Same reason as above: one function does setup and checks, the other one the actual work. I'll move it, too.

carlosgcampos commented on Source/WebKit/Platform/IPC/unix/UnixMessage.h:
```diff
@@ -75,15 +87,30 @@ class MessageInfo {
+    inline void clear()
+    {
+        memset(static_cast<void*>(this), 0, sizeof(MessageInfo));
+    }
```
Move this before the variables too.

carlosgcampos reviewed (approved):
Looks good, please check my comments.

spenap commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -490,11 +579,27 @@ bool Connection::sendOutputMessage(UnixMessage& outputMessage)
+                    if (!fd)
+                        return false;
+                    fdPtr[fdIndex++] = fd.value();
+                    return false;
```
```
                    if (!!fd) {
                        ASSERT(fdPtr);
                        fdPtr[fdIndex++] = fd.value();
                        // return true???
                    }
                    return false;
```

This adds an assert, to be consistent with [line 598](https://github.com/WebKit/WebKit/pull/49714/files#diff-19e014c216f6bfd1c11216423542cd7dab4e48502122fec35cf7b06cbfb65161R598), and compacts the condition since we seem to be returning `false` regardless.

That means that valid attachments are only the ones that carry valid buffers, right? Unless the intention was to return `true`, as `processMessage` seems to hint towards.

aperezdc replied:
Actually, this is a good catch 😱

This should be returning `true` here if the file descriptor is valid:

```suggestion
                    if (!fd)
                        return false;
                    fdPtr[fdIndex++] = fd.value();
                    return true;
```

Valid file descriptors need to be considered valid attachments as well!

spenap commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -192,11 +221,29 @@ bool Connection::processMessage()
+        switch (attachmentInfo[i].type()) {
+        case AttachmentInfo::Type::FileDescriptor:
+            attachments[attachmentCount - i - 1] = UnixFileDescriptor {
+                attachmentInfo[i].isNull() ? -1 : m_fileDescriptors[fdIndex++],
```
Hummm... but in `Connection::sendOutputMessage` we're setting the attachment info as null unconditionally for file descriptors, so wouldn't this always be `-1`?

spenap replied:
[Line 595](https://github.com/WebKit/WebKit/pull/49714/files/27cb3ab09e0b7327448d97b060bbba045ae69d5f#diff-19e014c216f6bfd1c11216423542cd7dab4e48502122fec35cf7b06cbfb65161R595) does `setNull` if invalid, and we set it to invalid in both options for file descriptor attachments

spenap commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -192,11 +221,29 @@ bool Connection::processMessage()
+        case AttachmentInfo::Type::FileDescriptor:
+            attachments[attachmentCount - i - 1] = UnixFileDescriptor {
+                attachmentInfo[i].isNull() ? -1 : m_fileDescriptors[fdIndex++],
+                UnixFileDescriptor::Adopt
```
...and here I am confused. If we use `Adopt`, that means this takes ownership, but on the `sending` side we did `value`, so we kept ownership -- we didn't do `release`.

aperezdc replied:
The receiving side gets a “new” file descriptor, that's why we `Adopt` here. The file descriptor table of the sender does not get changed.

According to the [unix(7)](https://www.man7.org/linux/man-pages/man7/unix.7.html) manual page, under the “Ancillary Messages” section (emphasis mine):

> Commonly, this operation is referred to as "passing a file descriptor" to another process.  However, more accurately, what is being passed is a reference to an open file description (see [open(2)](https://www.man7.org/linux/man-pages/man2/open.2.html)), and in the receiving process it is likely that a different file descriptor number will be used.  Semantically, **this operation is equivalent to duplicating ([dup(2)](https://www.man7.org/linux/man-pages/man2/dup.2.html)) a file descriptor** into the file descriptor table of another process.

carlosgcampos commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -179,11 +199,30 @@ bool Connection::processMessage()
     Vector<AttachmentInfo> attachmentInfo(attachmentCount);
 
+#if OS(ANDROID)
+    [[maybe_unused]] size_t validHardwareBufferCount = 0;
```
null is not necessarily invalid, I would use just attachmentHardwareBufferCount for consistency with attachmentFileDescriptorCount

aperezdc replied:
Makes sense, I'll rename this variable before landing; and also the corresponding `AttachmentInfo` members.  👍🏼

aperezdc committed:
[WPE][IPC] Add support for AHardwareBuffer attachments
https://bugs.webkit.org/show_bug.cgi?id=297592

Reviewed by Carlos Garcia Campos.

When building for Android, make message attachments a variant to
allow passing both UnixFileDescriptor or AHardwareBuffer through
IPC. This involved the following related changes in how Attachment
objects are handled: a new AttachmentInfo::m_type member indicates
whether an attachment is an UnixFileDescriptor or an AHardwareBuffer,
and alternative code paths are needed used to act them.

The trickier part of actually sending and receiving AHardwareBuffer
objects: we cannot "open up" the inside of buffer, and passing them
around needs to be done with AHardwareBuffer_sendHandleToUnixSocket()
and AHardwareBuffer_recvHandleFromUnixSocket(). These function will
themselves send/receive a message for each AHardwareBuffer, out of
our control. Therefore, on the writing end:

  - When preparing to send a message, note the amount of buffer
    handles in a new MessageInfo::m_validHardwareBufferCount
    member (invalid/null handles are not sent).

  - Store the valid AHardwareBuffer handles to be sent later and
    send the message (sans buffer handles) in the usual fashion.

  - Send the AHardwareBuffer handles right after the message.
    At the end of Connection::sendOutputMessage(), store the buffer
    handles Connection::m_outgoingHardwareBuffers, and call
    ::sendOutgoingHardwareBuffers().

  - The ::sendOutgoingHardwareBuffers() function attempts to send
    as many handles as it can right away. If the operation would
    block, it makes use of the socket monitor to re-schedule calling
    itself when the socket is ready again for sending. This is
    repeated until there are no more buffers left.

On the receiving end things are a little bit trickier because the
socket monitor is always armed to invoke Connection::readyReadHandler()
and this means that we cannot start it independently to handle only
retiring AHardwareBuffer handles from the socket -- we need to thread
the code to retire buffer handles into the existing functions:

  - The ::readyReadHandler() function may be invoked either to
    retry sending the "main" message, or to retry sending buffer
    handles. The Connection::m_pendingIncomingHardwareBufferCount
    is used to determine which code path to follow.

  - First, if there are no pending buffer handles to receive, a
    message is read. At the moment the code assumes that message
    are always read in their entirety, which is the case with
    SOCK_SEQPACKET.

  - The read message up to this point includes both the payload and
    file descriptors. The MessageInfo in the header of the message
    indicates how many AHardwareBuffer handles we need to retire.
    This is saved in Connection::m_pendingIncomingHardwareBufferCount
    before calling ::receiveIncomingHardwareBuffers().

  - Then ::receiveIncomingHardwareBuffers() attempts retiring from
    the socket as many buffer handles as it can. If reading handles
    would block, the function returns immediately, and will rely on
    the next call to ::readyReadHandler() -- as the socket monitor
    is always kept armed and invoke it for reading.

  - When execution reaches ::readyReadHandler() again and, crucially,
    there are still AHardwareBuffer handles pending to be retired from
    the socket, it will call again ::receiveIncomingHardwareBuffers()
    to resume the task. When all the handles have been retired from
    the socket we still need a call to ::processMessage() at this point,
    but without going first through trying to read another complete
    message.

* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/android/RefPtrAndroid.cpp: Added.
(WTF::DefaultRefDerefTraits<AHardwareBuffer>::refIfNotNull):
(WTF::DefaultRefDerefTraits<AHardwareBuffer>::derefIfNotNull):
* Source/WTF/wtf/android/RefPtrAndroid.h: Added.
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/Attachment.h:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/android/ArgumentCodersAndroid.cpp: Added.
(IPC::ArgumentCoder<UnixFileDescriptor>::decode):
(IPC::ArgumentCoder<RefPtr<AHardwareBuffer>>::encode):
(IPC::ArgumentCoder<RefPtr<AHardwareBuffer>>::decode):
* Source/WebKit/Platform/IPC/android/ArgumentCodersAndroid.h: Added.
* Source/WebKit/Platform/IPC/unix/ArgumentCodersUnix.cpp:
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::AttachmentInfo::AttachmentInfo):
(IPC::AttachmentInfo::type const):
(IPC::AttachmentInfo::setType):
(IPC::Connection::processMessage):
(IPC::Connection::readyReadHandler):
(IPC::Connection::platformCanSendOutgoingMessages const):
(IPC::Connection::sendOutputMessage):
(IPC::Connection::sendOutgoingHardwareBuffers):
(IPC::Connection::receiveIncomingHardwareBuffers):
* Source/WebKit/Platform/IPC/unix/UnixMessage.h:
(IPC::MessageInfo::MessageInfo):
(IPC::MessageInfo::hardwareBufferCount const):
(IPC::UnixMessage::UnixMessage):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesWPE.txt:

Canonical link: https://commits.webkit.org/299311@main
